src/jit/simdcodegenxarch.cpp

```diff
@@ -619,8 +619,8 @@ CodeGen::genSIMDIntrinsicInit(GenTreeSIMD* simdNode)
             // data section using movss/sd.  Similarly if op1 is a memory op we
             // would have loaded it using movss/sd.  Movss/sd when loading a xmm reg
             // from memory would zero-out upper bits. In these cases we can
-            // avoid explicitly zero'ing out targetReg.
-            bool zeroInitRequired = !(op1->IsCnsFltOrDbl() || op1->isMemoryOp());
+            // avoid explicitly zero'ing out targetReg if targetReg and op1Reg are the same.
+            bool zeroInitRequired = (targetReg != op1Reg) || !(op1->IsCnsFltOrDbl() || op1->isMemoryOp());
```

Comment:
Conceptually the way i look at it is that it is upto genSIMDScalarMove() to choose the right instruction when asked to zero initialize upper bytes of dst or preserve upper bytes of dst while copying or copy entire src to dst.

Even if we imagine there is a new instruction that zeros upper bytes, i would expect one of two things: caller of genSIMDScalarMove() will pass SMT_ZeroInitUpper if the src is not known to be initialized using such an instruction or will pass SMT_VerbatimCopy if caller can determine that src is initialized using such an instruction.

With SMT_CopyKnownUpperZeros/SMT_SrcHasUpperZeros we would need to pass two params to genSIMDScalarMove(): one to indicate whether to zero initialize or preserve upper bytes of dst and second whether src is known to have zeros in upper bytes already.  On the other hand if we still want a tri-state enum as single parameter, if I understood correctly, what you are suggesting as third enum is SMT_ZeroInitUpperWithSrcHasUpperZero and to me it is like combining two enums into one.